src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -0,0 +1,459 @@
+            out.write(str, off, len);
+            out.write(ANSI.showCursor());
+            out.flush();
+        }
+
+       
+    }
+
+    /**
+     * Represents one currently running progress bar
+     */
+    private class ProgressBar {
+        private final long threadId;
+        private final String threadName;
+        private final String name;
+        private int max;
+        private int current = 0;
+        private int previousWidth = 0;
+        private int doneWidth = 0;
+        private final int barWidth = lineWidth - "☐ ".length() - " 🕐 00:00:00.000 ".length();
+        private final Instant startTime;
+        private Duration duration;
+        private String message = "";
+        private int stepper = 1;
+        private final String[] clocks = new String[] {"🕐" , "🕑", "🕒", "🕓", "🕔", "🕕", "🕖", "🕗", "🕘", "🕙", "🕛"};
+        public int nesting = 0;
+
+        ProgressBar(String name, int max) {
+            this.threadId = Thread.currentThread().getId();
+            this.threadName = Thread.currentThread().getName();
+            this.name = name;
+            this.max = Math.max(1, max);
+            this.startTime = Instant.now();
+            this.duration = Duration.ZERO;
+            this.message = name;
+        }
+
+        void worked(int amount, String message) {
+            this.current += Math.min(amount, max);
+            this.duration = Duration.between(startTime, Instant.now());
+            this.message = message;
+        }
+
+        /**
+         * To avoid flickering of all bars at the same time, we only reprint
+         * the current bar
+         */
+        void update() {
+            // to avoid flicker we only print if there is a new bar character to draw
+            if (newWidth() != previousWidth) {
+                stepper++;
+                writer.write(ANSI.hideCursor());
+                writer.write(ANSI.moveUp(bars.size() - bars.indexOf(this)));
+                write();
+                writer.write(ANSI.moveDown(bars.size() - bars.indexOf(this) - 1 /* already wrote a \n */));
+                writer.write(ANSI.showCursor());
+                writer.flush();
+            }
+        }
+
+        int newWidth() {
+            if (max != 0) {
+                current = Math.min(max, current); // for robustness sake
+                var partDone = (current * 1.0) / (max * 1.0);
+                return (int) Math.floor(barWidth * partDone);
+            }
+            else {
+                return barWidth % stepper;
+            }
+        }
+
+        /**
+         * Print the current state of the progress bar
+         */
+        void write() {
+            previousWidth = doneWidth;
+            doneWidth = newWidth();
+                        
+            // var overWidth = barWidth - doneWidth;
+            var done = current >= max ? "☑ " : "☐ ";
+
+            // capitalize
+            var msg = message.substring(0, 1).toUpperCase() + message.substring(1, message.length());
```

Comment:
isn't this up to the user? just print the message as is, and let the user take care of proper casing of their use case.